src/main/java/ua/lviv/iot/searcher/Searcher.java

```diff
@@ -0,0 +1,35 @@
+package ua.lviv.iot.searcher;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import ua.lviv.iot.graph.Graph;
+
+public class Searcher {
+
+	public List<String> findOptimalSequenceFor(Graph<String> graph) {
+		int numberOfVertices = graph.numberOfVertices();
+		List<String> optimalSequence = new ArrayList<String>(numberOfVertices);
+		boolean[] isVisited = new boolean[numberOfVertices];
+		for (int counter = 0; counter < numberOfVertices; counter++) {
+			isVisited[counter] = false;
+		}
+		for (int i = 0; i < graph.numberOfVertices(); i++) {
+			depthFirstSearch(i, graph, optimalSequence, isVisited);
+		}
+		return optimalSequence;
+	}
+
+	private void depthFirstSearch(int currentVertexPosition, Graph<String> graph, List<String> optimalSequence,
```

Comment:
So... Why do you not return optimalSequence, but rather pass it as a param? What's the point? When you design your function syntax – think about an interface. If someone would be using it tomorrow – would they understand how to call this function?

This trick sometimes is done if you need to return several values and you don't want to create a class to incapsulate them – you just pass 2-3 vars as params and read their value afterwards. This is a bad habit though.

Reply:
Also, regarding `private`... Unless you absolutely know why – never use `private`, when you can use `protected`. Read up the difference and try to understand why I recommend protected

Reply:
I'm not sure about returning optimalSequence from depthFirstSearch. 

My logic was that I declare and initialize oprimalSequence only once in findOptimalSequenceFor function. Then I pass it as param to depthFirstSearch function which is recursive and it just adds elements to already existing list (optimalSequence).

If I create this list in depthFirstSearch function then with every its recursive launch a new list will be created and then appended to previous on the way back.

Maybe I'm wrong, but it doesn't seem to be optimal. Or there is another approach of implementing it so that optimalSequence is returned from depthFirstSearch?

Reply:
Right. There's always a balance that you need to look for between optimisation and usability: a simple example is web-site UI: optimal thing is to put a link performing specific action only once, but due to UX we sometimes create several links with the same action in different parts of the page. Have you got the analogy?

Now, if you want to optimise things here, there are several ways:
* Implement an iterative algorithm — this way you don't have to think about recursion issues
* Create 2 functions: one public and one private (protected in Java terms). Create good, user-oriented public interface and do whatever nasty thing you need in private one. Call private from public
* Don't bother optimising for object creation, unless it's huge. In reality, recursion with every call serializes tons of state and puts it to the call stack. So... Yet another car with short livespan should not impact your code much. There's GC to clean up for you (not a best mantra, but works in this case)